cartography/intel/semgrep/findings.py

```diff
@@ -0,0 +1,217 @@
+    """
+    all_vulns = []
+    sca_url = f"https://semgrep.dev/api/sca/deployments/{deployment_id}/vulns"
+    has_more = True
+    cursor = ""
+    prev_request = None
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": f"Bearer {semgrep_app_token}",
+    }
+
+    while has_more:
+        params = {}
+        if cursor:
+            params = {"cursor": cursor}
+        try:
+            response = requests.get(sca_url, params=params, headers=headers)
+            response.raise_for_status()
+            if prev_request == response.request.url:
+                logger.warning(
+                    "Duplicate request detected. Breaking the loop to avoid infinite requests.",
+                )
+                break
+            prev_request = response.request.url
+            if response.status_code == 200:
+                data = response.json()
+                vulns = data["vulns"]
+                cursor = data.get("cursor")
+                has_more = data.get("hasMore", False)
+                all_vulns.extend(vulns)
+        except requests.RequestException as e:
+            logger.error(
+                "Could not complete request to the SCA Semgrep API: %s", e, exc_info=e,
+            )
+            raise e
+        except Exception as e:
+            logger.error("Erorr retrieving Semgrep SCA vulns info: %s", e, exc_info=e)
+            raise e
+    return all_vulns
+
+
+def transform_sca_vulns(raw_vulns: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """
+    Transforms the raw SCA vulns response from Semgrep API into a list of dicts
+    that can be used to create the SemgrepSCAFinding nodes.
+    """
+    vulns = []
+    for vuln in raw_vulns:
+        sca_vuln: Dict[str, Any] = {}
+        # Mandatory fields
+        unique_id = f"{vuln['repositoryName']}|{vuln['advisory']['ruleId']}"
+        sca_vuln["id"] = unique_id
+        sca_vuln["repositoryName"] = vuln["repositoryName"]
+        sca_vuln["ruleId"] = vuln["advisory"]["ruleId"]
+        sca_vuln["title"] = vuln["advisory"]["title"]
+        sca_vuln["description"] = vuln["advisory"]["description"]
+        sca_vuln["ecosystem"] = vuln["advisory"]["ecosystem"]
+        sca_vuln["severity"] = vuln["advisory"]["severity"]
+        sca_vuln["reachability"] = vuln["advisory"]["reachability"]
+        sca_vuln["reachableIf"] = vuln["advisory"]["reachableIf"]
+        sca_vuln["exposureType"] = vuln["exposureType"]
+        dependency = f"{vuln['matchedDependency']['name']}|{vuln['matchedDependency']['versionSpecifier']}"
+        sca_vuln["matchedDependency"] = dependency
+        sca_vuln["dependencyFileLocation_path"] = vuln["dependencyFileLocation"]["path"]
+        sca_vuln["dependencyFileLocation_url"] = vuln["dependencyFileLocation"]["url"]
+        # Optional fields
+        cves = vuln.get("advisory", {}).get("references", {}).get("cveIds")
+        if len(cves) > 0:
+            # Take the first CVE
```

Comment:
Why only the first cve?

Reply:
From the tests I ran, it returns either 0 or 1 CVE that array.
Having a single CVE is useful because we want to match other scanner results 1 to 1.
In the future maybe we can use the CVE module https://github.com/lyft/cartography/tree/master/cartography/intel/cve, but that would lead to more complex relationships in order to connect a CVE from other scanner to each Semgrep finding.

Reply:
This endpoint is not yet documented on their public docs: https://semgrep.dev/api/v1/docs/#section/Introduction. I assume they made it an array for future cases.

The way we intend to use this is 1 SCA finding = 1 vuln at most (there are cases where there is no CVE).
Adding multiple CVEs to each finding will then cause us to search
```
MATCH (other:OtherCVEScanner)
MATCH (sca:SCAFinding)-[:FINDING_TO_CVE]-(cve:SCACVE)
WHERE other.id = cve.id
```
Considering there aren't cases yet where `FINDING_TO_CVE` would be more than 1, I thing we can save the complexity of adding an extra relationship.

Reply:
Adding a new relationship is super easy with the data model now ;)

Once we define a "canonical" CVE node in the data model, it'll make sense to include it but I see your point.